### DISCUSSION TITLE
How to call a handler from a task expression

### DISCUSSION THREAD

**Discussion by s5no5t** [giraffe-fsharp/Giraffe#536] [Q&A]
I'm trying to build a handler that opens and closes a document database (RavenDB) session, and calls other handlers in between.

```fsharp
let handler session: HttpHandler =
    choose
        [ other handlers.. ]

let sessionHandler documentStore: HttpHandler =
    fun next ctx ->
        task {
            let session = RavenDb.createSession documentStore
            let result = (handler session) next ctx // <--- what's the correct way to call handler here?
            do! RavenDb.saveChangesAsync session
            return! result
        }

let webApp documentStore = HttpHandlers.sessionHandler documentStore

```

However, when I try the above code, the database complains about `saveChangedAsync` being called concurrently:

```
System.InvalidOperationException: Concurrent usage of async tasks in async session is forbidden, please make sure that async session doesn't execute async methods concurrently.
```

Any idea how to call other handlers and properly "await" the result before continuing in the task expression?

**Comment by 64J0** [comment]
It could be the case that `saveChangeAsync` is returning an **async** operation, which is different from a **task** operation in .NET @s5no5t. In this case, you need to convert the **async** to **task** to use it properly in this CE.

I guess you'll need to do this:

```fsharp
do! (RavenDb.saveChangesAsync session) |> Async.StartAsTask
```

More information here -> https://learn.microsoft.com/en-us/dotnet/fsharp/tutorials/async#core-concepts.

**Reply by s5no5t** [reply]
In this case, I'm sure it's returning a `Task`, as it's a method from a [C# library](https://ravendb.net/docs/article-page/5.4/csharp/client-api/session/saving-changes)

**Reply by 64J0** [reply]
Got it, thanks for the reference!

**Comment by TheJayMann** [comment]
I'm not entirely sure as I have not thoroughly evaluated the provided code, nor have I tested anything, but I believe it could be fixed by moving the `bang` operation from the return to the let, as follows.
```fsharp
 let handler session: HttpHandler =
    choose
        [ other handlers.. ]

let sessionHandler documentStore: HttpHandler =
    fun next ctx ->
        task {
            let session = RavenDb.createSession documentStore
            let! result = (handler session) next ctx // use `let!` to await the result
            do! RavenDb.saveChangesAsync session
            return result  // use a regular return, as the result has been previously awaited
        }

let webApp documentStore = HttpHandlers.sessionHandler documentStore
```
